File: src/main/scala/org/jglr/phiengine/network/channels/NetworkChannel.java
package org.jglr.phiengine.network.channels;

import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandler;
import org.jglr.phiengine.network.NetworkSide;
import org.jglr.phiengine.core.PhiEngine;
import org.jglr.phiengine.network.Message;
import org.jglr.phiengine.network.Packet;
import org.jglr.phiengine.network.PacketHandler;

public class NetworkChannel implements ChannelInboundHandler {

    private final String name;
    private final NetworkSide side;
    private ChannelHandlerContext context;

    public NetworkChannel(String name, NetworkSide side) {
        this.name = name;
        this.side = side;
    }

    public String getName() {
        return name;
    }

    public ChannelHandlerContext getContext() {
        return context;
    }

    public void write(Packet packet) {
        packet.setChannel(name);
        context.write(packet); // TODO: Add listener
    }

    public void flush() {
        context.flush();
    }

    public void onConnection() {

    }

    @Override
    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
        this.context = ctx;
    }

    @Override
    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {

    }

    @Override
    public void channelActive(ChannelHandlerContext ctx) throws Exception {
        onConnection();
    }

    @Override
    public void channelInactive(ChannelHandlerContext ctx) throws Exception {

    }

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
        Message message = (Message)msg;
        Packet packet = message.createPacket();
        packet.read(message.payload);
        if(message.getSide() == side) {
            PacketHandler handler = PhiEngine.getInstance().getNetworkHandler().getHandler(packet.getClass());
            if (side.isClient()) {
                handler.handleClient(packet);
            } else {
                handler.handleServer(packet);
            }
        }
    }

    @Override
    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {

    }

    @Override
    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {

    }

    @Override
    public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {

    }

    @Override
    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {

    }

    @Override
    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {

    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
        PhiEngine.crash("Error while reading network", cause);
    }
}

File: src/main/scala/org/jglr/phiengine/network/channels/NetworkChannel.scala
package org.jglr.phiengine.network.channels

import io.netty.channel.{ChannelFuture, ChannelFutureListener, ChannelHandlerContext, ChannelInboundHandler}
import io.netty.util.concurrent.GenericFutureListener
import org.jglr.phiengine.core.PhiEngine
import org.jglr.phiengine.network.NetworkSide.NetworkSide
import org.jglr.phiengine.network.{PacketHandler, Message, Packet}

class NetworkChannel(private val name: String, private val side: NetworkSide) extends ChannelInboundHandler {
  private var context: ChannelHandlerContext = null

  def getName: String = {
    name
  }

  def getContext: ChannelHandlerContext = {
    context
  }

  def writeFlush(packet: Packet): Unit = {
    write(packet)
    flush()
  }

  def write(packet: Packet): Unit = {
    packet.setChannel(name)
    //.addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE)
    context.write(packet).addListener(new GenericFutureListener[ChannelFuture] {
      override def operationComplete(future: ChannelFuture): Unit = {
        if(future.isDone) {
          println(": "+future.isSuccess)
          if(future.cause() != null) {
            future.cause().printStackTrace()
          }
        }
      }
    })
    println("write")
  }

  def flush(): Unit = {
    context.flush
  }

  def onConnection(): Unit = {
  }

  @throws(classOf[Exception])
  def channelRegistered(ctx: ChannelHandlerContext) {
  }

  @throws(classOf[Exception])
  def channelUnregistered(ctx: ChannelHandlerContext): Unit = {
  }

  @throws(classOf[Exception])
  def channelActive(ctx: ChannelHandlerContext) {
    this.context = ctx
    onConnection()
  }

  @throws(classOf[Exception])
  def channelInactive(ctx: ChannelHandlerContext): Unit = {
  }

  @throws(classOf[Exception])
  def channelRead(ctx: ChannelHandlerContext, msg: AnyRef) {
    println("channelRead!! :DDD")
    val message: Message = msg.asInstanceOf[Message]
    val packet: Packet = message.createPacket
    packet.read(message.payload)
    if (message.getSide == side) {
      val handler: PacketHandler[Packet] = PhiEngine.getInstance.getNetworkHandler.getHandler(packet.getClass).asInstanceOf[PacketHandler[Packet]]
      if (side.isClient) {
        handler.handleClient(packet)
      } else {
        handler.handleServer(packet)
      }
    }
  }

  @throws(classOf[Exception])
  def channelReadComplete(ctx: ChannelHandlerContext) {
  }

  @throws(classOf[Exception])
  def userEventTriggered(ctx: ChannelHandlerContext, evt: AnyRef) {
  }

  @throws(classOf[Exception])
  def channelWritabilityChanged(ctx: ChannelHandlerContext) {
  }

  @throws(classOf[Exception])
  def handlerAdded(ctx: ChannelHandlerContext) {
  }

  @throws(classOf[Exception])
  def handlerRemoved(ctx: ChannelHandlerContext) {
  }

  @throws(classOf[Exception])
  def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
    PhiEngine.crash("Error while reading network", cause)
  }
}